en/05_Contributing/05_Coding_Conventions/01_PHP_Coding_Conventions.md

```diff
@@ -136,6 +136,133 @@ class MyClass extends Class
+
+## Interfaces
+
+- If there are or there's a realistic expection that there will be multiple class implementations of something then create an interface and use that for parameter and return types, and as the service key for injector.
+- If there is only a single implemention and no forseeable expectation of more implementaitons being added, then do not create an interface for the single class implementation because this would only be adding unecessarily complexity.
+
+## Typing
+
+- Aim to strongly type all new properties, method parameters, and method return types.
+- The exception to this is adding a new method or property to a class which is entirely untyped and adding typed API would be very inconsistent with the rest of the file. In this instance use PHP docblocks to define types.
+- If there are duplicate strong types and docblock types then remove the old docblock types.
+- Do not strongly type existing API outside of a new major version as it's an API breakage.
+
+## Mixed types
+
+- Where possible avoid using mixed types, including type unions, because this leads to complicated and brittle code.
+- Try to avoid using nullable scalars and instead use the default value of the scalar instead of null. One exception to this is for when `null` means 'uninitialised`.
+- Avoid the use of the splat `...` parameter as this makes API enforcement difficult. If you need to support multiple arguments then use an array parameter type.
+
+## Object instantiation
+
+- Use Injector to instantiate objects rather than the `new` keyword. This is done as it allows projects to swap out classes with different implementations.
+- Note there is only a minimal performance penalty for this, though there is a downside that IDE type hinting stops working. During development you may wish to use the new keyword for this reason and change it to use injector before submitting your PR.
+- For classes that use the Injectable trait use the `::create()` static method.
+- For objects that do not have the Injectable trait, use `Injector::inst()->get({classname})`
+- This also includes instantiation of third party classes `<<< @todo why? discuss`
+- The exception to this is for unit-tests where you should use the `new` because there you need to ensure you're testing a specific class
+
+## Extensions and traits
+
+- Use the DataExtension class for extending DataObjects.
+- Use the Extension class for extending non-DataObjects.
+- Typically you shouldn't use native PHP traits. A common gotcha is mistakenly using a trait on a DataObject to add a private static config property. This doesn't work because private static config properties are a Silverstripe specific thing and the Silverstripe DataExtension will correctly merge them. However traits are unable to merge private static config properites.
+
+### Extension hooks
+
+- Naming convention for extension hooks is `update{methodName}` where methodName has any leading `get` omitted has the remaining first letter uppercased.
```

Comment:
```suggestion
- Naming convention for extension hooks is `update{methodName}` where methodName has any leading `get` omitted and has the remaining first letter uppercased.
```

Reply:
That first point should probably even be more generic than that, and be "any time an extension hook is intended to updated a value" instead of just for getter methods. E.g. `updateDeleteTables()` in https://github.com/silverstripe/silverstripe-framework/blob/8680b6bfb93235547bfac29fcbd9edc8b13c7ab9/src/ORM/DataObject.php#L1752

Reply:
> We could also do away with the "on" part 

I think the "on" does have some merit as it makes it more recognizable as an extension hook, similar to the `get` prefix discussion elsewhere in this PR to make accessors more obvious what they are, and the 'update' prefix for most other hooks.

I think this may be more important if the hook is added via `$this->invokeWithExtensions('onBeforeSendResponse', $obj)` vs regular `$this->extend('onBeforeSendResponse', $obj);` because the former also allows the hooks to be implemented directly on subclasses rather than only on extensions.  In that scenario it's more obvious that `function onBeforeSendResponse(&$obj)' is a hook vs `function beforeSendResponse(&$obj)';

Reply:
Have rewritten to make it explicit there are two hook scenarios:
- `update` hooks that update the value before it's returned from a method
- `onBefore/onAfter` hooks that update a value before or after calling another method within the method